File: pysrc/medium2.cpp
#include "medium2.h"
#include <memory>
#include <pybind11/pybind11.h>
#include <pybind11/operators.h>
#include <pybind11/stl.h>

namespace py = pybind11;
using namespace pybind11::literals;

namespace batoid {
    void pyExportMedium2(py::module& m) {
        py::class_<Medium2>(m, "CPPMedium2")
            .def("getN", &Medium2::getN)
            .def("getNMany",
                [](const Medium2& m, size_t n, size_t wavelength_ptr, size_t out_ptr)
                {
                    m.getNMany(
                        reinterpret_cast<double*>(wavelength_ptr),
                        n,
                        reinterpret_cast<double*>(out_ptr)
                    );
                }
            );


        py::class_<ConstMedium2, Medium2>(m, "CPPConstMedium2")
            .def(py::init<double>());


        py::class_<SellmeierMedium2, Medium2>(m, "CPPSellmeierMedium2")
            .def(py::init<double,double,double,double,double,double>());


        py::class_<SumitaMedium2, Medium2>(m, "CPPSumitaMedium2")
            .def(py::init<double,double,double,double,double,double>());


        py::class_<Air2, Medium2>(m, "CPPAir2")
            .def(py::init<double,double,double>());
    }
}

File: batoid/medium2.py
import numpy as np

from . import _batoid

class Medium2:
    def getN(self, wavelength):
        import numbers
        if isinstance(wavelength, numbers.Real):
            return self._medium.getN(wavelength)
        else:
            wavelength = np.ascontiguousarray(wavelength)
            out = np.empty_like(wavelength)
            self._medium.getNMany(
                len(wavelength),
                wavelength.ctypes.data,
                out.ctypes.data
            )
            return out


class ConstMedium2(Medium2):
    def __init__(self, n):
        self._medium = _batoid.CPPConstMedium2(n)


# class TableMedium(Medium):
#     def __init__(self, table):
#         self.table = table
#         self._medium = _batoid.CPPTableMedium(self.table._table)
#
#     def __repr__(self):
#         return "TableMedium({!r})".format(self.table)


class SellmeierMedium2(Medium2):
    def __init__(self, B1, B2, B3, C1, C2, C3):
        self._medium = _batoid.CPPSellmeierMedium2(B1, B2, B3, C1, C2, C3)


class SumitaMedium2(Medium2):
    def __init__(self, A0, A1, A2, A3, A4, A5):
        self._medium = _batoid.CPPSumitaMedium2(A0, A1, A2, A3, A4, A5)


class Air2(Medium2):
    def __init__(self, pressure=69.328, temperature=293.15, h2o_pressure=1.067):
        self._medium = _batoid.CPPAir2(pressure, temperature, h2o_pressure)
